File: go/codec/codec.go
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

// Package codec provides the GTS transform catalog decoder (§8).
//
// It intentionally implements the baseline decode path only: identity, gzip,
// zstd, and zstd-rsyncable are decoded locally, while encrypt-class entries
// degrade to missing-key unless a higher layer supplies key-aware handling.
package codec

import (
	"bytes"
	"fmt"
	"io"

	"github.com/klauspost/compress/gzip"
	"github.com/klauspost/compress/zstd"
)

const maxZstdDecodedSize = 16 * 1024 * 1024

// Codec is a catalog entry (§5, §8.5).
type Codec struct {
	// Name is the canonical codec name looked up from the segment catalog.
	Name string
	// Cls is "encode", "compress", or "encrypt".
	Cls string
}

// Error describes why a transform chain could not be reversed.
type Error struct {
	// Reason is "unknown-codec" or "missing-key" for opaque degradation.
	Reason string
	// Detail is a human-readable diagnostic.
	Detail string
	// Failed is true when the codec is known but the bytes are corrupt.
	Failed bool
}

func (e *Error) Error() string {
	return e.Detail
}

// decodeOne reverses a single codec entry on data (identity, gzip, or zstd).
// Encrypt-class codecs cannot be reversed without a key.
func decodeOne(codec *Codec, data []byte) ([]byte, error) {
	if codec == nil {
		return nil, &Error{Failed: true, Detail: "codec chain contains nil entry"}
	}
	if codec.Cls == "encrypt" {
		return nil, &Error{
			Reason: "missing-key",
			Detail: fmt.Sprintf("no key for encrypt codec '%s'", codec.Name),
		}
	}
	switch codec.Name {
	case "identity":
		return data, nil
	case "gzip":
		r, err := gzip.NewReader(bytes.NewReader(data))
		if err != nil {
			return nil, &Error{Failed: true, Detail: fmt.Sprintf("gzip decode failed: %v", err)}
		}
		out, err := io.ReadAll(r)
		_ = r.Close()
		if err != nil {
			return nil, &Error{Failed: true, Detail: fmt.Sprintf("gzip decode failed: %v", err)}
		}
		return out, nil
	case "zstd", "zstd-rsyncable":
		r, err := zstd.NewReader(bytes.NewReader(data))
		if err != nil {
			return nil, &Error{Failed: true, Detail: fmt.Sprintf("zstd decoder init failed: %v", err)}
		}
		defer r.Close()
		out, err := io.ReadAll(io.LimitReader(r, int64(maxZstdDecodedSize+1)))
		if err != nil {
			return nil, &Error{Failed: true, Detail: fmt.Sprintf("zstd decode failed: %v", err)}
		}
		if len(out) > maxZstdDecodedSize {
			return nil, &Error{Failed: true, Detail: "zstd decode failed: decompressed size exceeds safety bound"}
		}
		return out, nil
	default:
		return nil, &Error{
			Reason: "unknown-codec",
			Detail: fmt.Sprintf("unknown codec '%s'", codec.Name),
		}
	}
}

// DecodeChain reverses a resolved codec chain, last to first (§6.1, §8.2).
//
// The baseline carries no keys, so every encrypt-class codec degrades to
// missing-key (matching the Python reader with keys=None).
func DecodeChain(chain []*Codec, data []byte) ([]byte, error) {
	current := data
	for i := len(chain) - 1; i >= 0; i-- {
		if chain[i] == nil {
			return nil, &Error{Failed: true, Detail: fmt.Sprintf("codec chain contains nil entry at index %d", i)}
		}
		var err error
		current, err = decodeOne(chain[i], current)
		if err != nil {
			return nil, err
		}
	}
	return current, nil
}

File: kotlin/src/main/kotlin/ca/blackcatinformatics/gts/Codec.kt
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

package ca.blackcatinformatics.gts

import com.github.luben.zstd.Zstd
import com.github.luben.zstd.ZstdInputStream
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

data class Codec(val name: String, val cls: String)

class CodecException(
    message: String,
    val reason: String,
    val failed: Boolean,
    cause: Throwable? = null,
) : RuntimeException(message, cause)

fun decodeChain(chain: List<Codec>, payload: ByteArray): ByteArray {
    var out = payload
    for (codec in chain.asReversed()) {
        out =
            when (codec.name) {
                "identity" -> out
                "gzip" -> gunzip(out)
                "zstd", "zstd-rsyncable" -> zstdDecompress(out)
                "cose-encrypt0" -> throw CodecException("missing content key for cose-encrypt0", "missing-key", false)
                else -> throw CodecException("unsupported codec ${codec.name}", "unknown-codec", false)
            }
    }
    return out
}

fun encodeChain(chain: List<Codec>, payload: ByteArray): ByteArray {
    var out = payload
    for (codec in chain) {
        out =
            when (codec.name) {
                "identity" -> out
                "gzip" -> gzip(out)
                "zstd", "zstd-rsyncable" -> Zstd.compress(out)
                else -> throw CodecException("unsupported codec ${codec.name}", "unknown-codec", true)
            }
    }
    return out
}

private fun gzip(data: ByteArray): ByteArray {
    val out = ByteArrayOutputStream()
    GZIPOutputStream(out).use { it.write(data) }
    return out.toByteArray()
}

private fun gunzip(data: ByteArray): ByteArray =
    try {
        GZIPInputStream(ByteArrayInputStream(data)).use { input ->
            val out = ByteArrayOutputStream()
            val buf = ByteArray(8192)
            while (true) {
                val n = input.read(buf)
                if (n < 0) break
                out.write(buf, 0, n)
            }
            out.toByteArray()
        }
    } catch (err: IOException) {
        throw CodecException("gzip decode failed: ${err.message}", "damaged", true, err)
    }

private fun zstdDecompress(data: ByteArray): ByteArray =
    try {
        ZstdInputStream(ByteArrayInputStream(data)).use { input ->
            val out = ByteArrayOutputStream()
            val buf = ByteArray(8192)
            while (true) {
                val n = input.read(buf)
                if (n < 0) break
                out.write(buf, 0, n)
            }
            out.toByteArray()
        }
    } catch (err: IOException) {
        throw CodecException("zstd decode failed: ${err.message}", "damaged", true, err)
    }
